### PR TITLE
Update HashTreeRoot v0.6

### PR DESCRIPTION
**Pull request by rauljordan**
This is part of #3 and a follow-up to #5, which implements Merkleization dependent on type as defined in the SSZ spec as follows:

Define hash_tree_root(value) of an object value recursively:

- `merkleize(pack(value))` if value is a basic object or a vector of basic objects
- `mix_in_length(merkleize(pack(value)), len(value))` if value is a list of basic objects
- `merkleize([hash_tree_root(element) for element in value])` if value is a vector of composite objects or a container
- `mix_in_length(merkleize([hash_tree_root(element) for element in value]), len(value))` if value is a list of composite objects
- `mix_in_type(merkleize(value.value), value.type_index)` if value is of union type

### CODE REVIEW

File: utils.go [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+	copy(y[:], x)
+	return y
+}
+
+// Hash defines a function that returns the
+// Keccak-256/SHA3 hash of the data passed in.
+// https://github.com/ethereum/eth2.0-specs/blob/master/specs/core/0_beacon-chain.md#appendix
+func Hash(data []byte) [32]byte {
+	var hash [32]byte
+	h := sha3.NewLegacyKeccak256()
```
Comment (nisdas) [inline utils.go]: change this to sha256

File: tree_hash.go [hunk 2]
```diff
@@ -0,0 +1,267 @@
+package ssz
```
Comment (rauljordan) [inline tree_hash.go]: Review this file

File: tree_hash.go [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"reflect"
+)
+
+const hashLengthBytes = 32
```
Comment (prestonvanloon) [inline tree_hash.go]: Can you add a comment what this references in the spec? I think it's `BYTES_PER_CHUNK`.

File: tree_hash.go [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+const sszChunkSize = 128
```
Comment (prestonvanloon) [inline tree_hash.go]: Where does this number come from? Can't find in spec either.

File: tree_hash.go [hunk 5]
```diff
@@ -0,0 +1,267 @@
...
+	return hasher, nil
+}
+
+func makePtrHasher(typ reflect.Type) (hasher, error) {
+	elemSSZUtils, err := cachedSSZUtilsNoAcquireLock(typ.Elem())
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO(1461): The tree-hash of nil pointer isn't defined in the spec.
```
Comment (prestonvanloon) [inline tree_hash.go]: What issue does this reference? Something in Prysm? We should transfer it to this repo and use the new issue number if that is the case.

File: helpers.go [hunk 6]
```diff
@@ -1,14 +1,22 @@
...
 // chunks, right-pad the last chunk with zero bytes, and return the chunks.
 // Basic types are either bool, or uintN where N = {8, 16, 32, 64, 128, 256}.
 //
 // Important: due to limitations in Go generics, we will assume the input is already
 // a list of SSZ-encoded objects of the same type.
 func pack(serializedItems [][]byte) ([][]byte, error) {
+	areAllEmpty := true
+	for _, item := range serializedItems {
+		if !bytes.Equal(item, []byte{}) {
+			areAllEmpty = false
```
Comment (prestonvanloon) [inline helpers.go]: break?

File: hash_cache.go [hunk 7]
```diff
@@ -109,31 +107,28 @@ func (b *hashCacheS) TrieRootCached(val interface{}) ([32]byte, error) {
 // least recently added root info if the cache size has reached the max cache
 // size limit.
 func (b *hashCacheS) MerkleHashCached(byteSlice [][]byte) ([]byte, error) {
```
Comment (prestonvanloon) [inline hash_cache.go]: Why not return `[32]byte`?

File: hash_tree_root.go [hunk 8]
```diff
@@ -0,0 +1,256 @@
...
+func isBasicType(kind reflect.Kind) bool {
+	return kind == reflect.Bool ||
+		kind == reflect.Uint8 ||
+		kind == reflect.Uint16 ||
+		kind == reflect.Uint32 ||
+		kind == reflect.Uint64
+}
+
+func isBasicTypeArray(typ reflect.Type, kind reflect.Kind) bool {
+	return kind == reflect.Array && typ.Elem().Kind() == reflect.Bool ||
```
Comment (prestonvanloon) [inline hash_tree_root.go]: return kind == reflect.Array && isBasicType(type.Elem.Kind())

File: hash_tree_root.go [hunk 9]
```diff
@@ -0,0 +1,256 @@
...
+		kind == reflect.Array && typ.Elem().Kind() == reflect.Uint8 ||
+		kind == reflect.Array && typ.Elem().Kind() == reflect.Uint16 ||
+		kind == reflect.Array && typ.Elem().Kind() == reflect.Uint32 ||
+		kind == reflect.Array && typ.Elem().Kind() == reflect.Uint64
+}
+
+func isBasicTypeSlice(typ reflect.Type, kind reflect.Kind) bool {
+	return kind == reflect.Slice && typ.Elem().Kind() == reflect.Bool ||
```
Comment (prestonvanloon) [inline hash_tree_root.go]: return kind == reflect.Slice && isBasicType(type.Elem.Kind())

File: utils.go [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+// sized 32 byte array. This method will truncate the input if it is larger
+// than 32 bytes.
+func ToBytes32(x []byte) [32]byte {
+	var y [32]byte
+	copy(y[:], x)
+	return y
+}
+
+// Hash defines a function that returns the sha256 hash of the data passed in.
+// https://github.com/ethereum/eth2.0-specs/blob/master/specs/core/0_beacon-chain.md#appendix
```
Comment (prestonvanloon) [inline utils.go]: Please reference a specific commit or this link will decay

File: helpers.go (hunk 6 above)
Comment (rauljordan) [reply helpers.go]: This tests if all are empty, not that is any empty

File: hash_cache.go (hunk 7 above)
Comment (rauljordan) [reply hash_cache.go]: Hash cache isn't handled by this PR, I think it'll require another PR to update and ensure ti works

File: helpers.go (hunk 6 above)
Comment (prestonvanloon) [reply helpers.go]: Please add the break. As soon as you find one that is not empty, you can stop checking.

**Pull request merged**